src/computability/regular_expressions.lean

```diff
@@ -0,0 +1,295 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Fox Thomson.
+-/
+import data.fintype.basic
+import data.finset.basic
+import tactic.rcases
+import tactic.omega
+
+/-!
+# Regular Expressions
+
+This file contains the formal definition for regular expressions and basic lemmas. Note these are
+regular expressions in terms of formal language theory. Note this is different to regex's used in
+computer science such as the POSIX standard.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- This is the definition of regular expressions. The names used here is to mirror the definition
+  of a Kleene algebra (https://en.wikipedia.org/wiki/Kleene_algebra).
+  `RZero` matches nothing
+  `RNull` matches only the empty string
+  `RChar a` matches only the string 'a'
+  `RStar M` matches any finite concatenation of strings which match `M`
+  `RPlus M N` matches anything which match `M` or `N`
+  `RComp M N` matches `x ++ y` if `x` matches `M` and `y` matches `N` -/
+inductive regular_expression (α : Type u) : Type (u+1)
+| RZero : regular_expression
```

Comment:
```suggestion
| zero : regular_expression
```
Constructors are typically lowercase and do not repeat the type.

I'd also use `epsilon` instead of `null`, at least that's more familiar to me.